Add a threat model to Tock's documentation. (#1632)
tock/tock: doc/threat_model/Untrusted_Capsule_Isolation.md

Untrusted Capsule Isolation
===========================

## Isolation Mechanism

> Untrusted capsules are limited to what they should be able to access within

Reviewer: ```suggestion
Untrusted capsules are limited to what they can access within
```

Author: Done.

Rust's type system without using `unsafe`. That isolation is implemented by

> banning `unsafe` from use in untrusted capsule code. This isolation is

Reviewer: ```suggestion
Rust's type system without using `unsafe`. That isolation is implemented by
banning `unsafe` from use in untrusted capsule code and any libraries the untrusted capsule
uses (except for the Rust core library). However, there is currently no automated mechanism
for enforcing this restriction in libraries, and therefore Tock prohibits using any external libraries.
This isolation is
```

Author: I didn't realize that untrusted capsules were limited in what libraries they can use. I adjusted this paragraph to indicate that limitation.

vulnerable to code that exploits compiler bugs or bugs in `unsafe` code in

> libraries. When a board integrator chooses to use an untrusted capsule, they are

Reviewer: ```suggestion
vulnerable to code that exploits compiler bugs.
When a board integrator chooses to use an untrusted capsule, they are
```

Author: This isolation is vulnerable to exploits in `libcore`'s `unsafe` sections.

responsible for auditing the code of the untrusted capsule to detect potentially

> malicious behavior. This relies in part on Rust's resistance to underhanded

Reviewer: What is "this"?

Author: I clarified it. I was referring to the use of Rust's type system as a security isolation mechanism.

programming techniques (stealthy obfuscation), and is a weaker form of isolation
than the hardware-backed isolation used to isolate the kernel (and other
applications) from applications.

## Impact on Kernel API Design

Kernel APIs should be designed to limit the data that untrusted capsules have
access to. Trusted kernel code should use capabilities as necessary in its API
to limit the access that untrusted capsule code has. For example, an API that
would allow an untrusted capsule to access data external to it should require a

> "trusted" capability.

Reviewer 2: I don't think that we currently meet this requirement. Technically capsules like virtual_spi have the ability to access all sorts of external data (all radio data for example) but do not require a capability (they instead require a reference to the chip Spi driver). I think we should probably provide a definition for what a capability is in the context of Tock, and elaborate that either a capability or an exclusive reference to some form of privileged object is required to limit such access. Examples would also probably be useful here.

Author: I clarified the wording a bit -- does it match your expectation now? I thought it already worked in the way I described.

Reviewer 2: I like the new wording, but I'm still not sure that Tock currently works this way. The `virtual_spi` capsule can access the data passed by any other capsule that uses SPI (i.e. the contents of a radio packet received over SPI or the value of a sensor reading completed over SPI), and the `virtual_spi` capsule does not require a "trusted" capability (which I interpret to mean a `XCapability` struct that requires `unsafe` to create).

Unless all data passed over SPI is not considered external to the `virtual_spi` capsule?

Author: Data passed over SPI is not considered external. Clients of the `virtual_spi` capsule would not be surprised that the `virtual_spi` capsule can see the data passed over SPI.

I'm not sure how to express that more precisely. My conceptual model is "well it's kinda obvious from the interface that the data will flow through that component" but it's hard to formalize.

> Untrusted Capsule Isolation

Reviewer 3: Should we be saying something about availability here? Currently manual  inspection is required to ensure that capsules do not destroy availability guarantees (maliciously or accidentally) through the use of `panic`, infinite loops, etc.

Reviewer 4: I think there's already an explicit assumption in the capsule design that they're scheduled cooperatively, from before the security model. But it's certainly should be mentioned in the document.

Author: Yes, I agree that this doc should discuss availability. Done.
